src/sml.cpp

```diff
@@ -289,28 +310,31 @@ void pow(double &val, signed char &scaler)
 void smlOBISByUnit(long long int &val, signed char &scaler, sml_units_t unit)
 {
   unsigned char i = 0, pos = 0, size = 0, y = 0;
+  sml_states_t type;
   val = -1; /* unknown or error */
   while (i < listPos) {
     pos++;
-    if (pos == 4 && listBuffer[i + 1] != unit) {
-      /* if unit at position 4 is not 0x1e (wh) return unknown */
+    size = (int)listBuffer[i++];
+    type = (sml_states_t)listBuffer[i++];
+    if (pos == 4 && listBuffer[i] != unit) {
+      /* return unknown (-1) if unit does not match */
       return;
     }
     if (pos == 5) {
-      scaler = listBuffer[i + 1];
+      scaler = listBuffer[i];
     }
     if (pos == 6) {
-      size = (int)listBuffer[i];
       val = 0;
+      y = size;
+      if (type == SML_DATA_SIGNED_INT && (listBuffer[i] & (1 << 7))) {
+        // negative value
+        val = 0xFFFFFFFFFFFFFFFF;
```

Comment:
NICE! This really makes sign extension simple.

I would suggest a small improvement, thoug:
```suggestion
        val = ~0;
```

This should give an all-ones value even if at some point in time the data type of val changes (again :)).